src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -231,24 +231,23 @@ boolean isOpen() {
 							RelayLogger.logEvent("writingBytesFinished", this, String.valueOf(text.length()));
 						}
 						else {
-							ByteBuffer bytes = null;
+							byte[] bytes;
 							if (data instanceof byte[]) {
-								bytes = ByteBuffer.wrap((byte[]) data);
+								bytes = ((byte[]) data).clone();
 							} 
 							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
-							}
-							else if (data instanceof String) {
-								bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+								ByteBuffer buffer = (ByteBuffer) data;
+								bytes = new byte[buffer.remaining()];
+								buffer.get(bytes);
```

Comment:
Sweet. Maybe ad a comment here about whatever buffer we pass being messed up by the remote.sendBinary call? Otherwise some well-meaning individual may "optimize" this extra copy out in the future.

Reply:
Oh, it's already there lower in this method.

---
In reply to: [296943483](https://github.com/Azure/azure-relay-java/pull/48#discussion_r296943483) [](ancestors = 296943483)